### DISCUSSION TITLE
Query a large collection Index with TimeSeries and how to optimize things to perform fast

### DISCUSSION THREAD

**Discussion by Welliba-Sven-Barkhahn** [ravendb/ravendb#12886] [Q&A]
Hello,

I'm fairly new to RavenDB and am in the process of figuring out if RavenDB meets my needs.

**Situation:**
A user can track his healthiness on a daily basis (ideally) and our service would store 'values' (simple integer) into a specific time series. It could be that a user updates many values per day or only a few per week. In my opinion, the time series fits well to store this data and we can use them later to virtualize it in a dashboard.

Users can have multiple properties and for simplicity I will use an Array that contains numbers. Let's call it Labels and it contains a Label_Id.

**What I want:**
I want to query users by an combination of multiple labels and do some calculations with the data in the time series. Getting this data should be fast and ideally under 2 seconds. It will be a large collection.

**What I did:**
I have a collection of 1.07 million documents (dummy data, just to check performance). Each Document has two timeSeries, where an integer is stored as a value.

Example of a Subject:
``` JSON
{
    "Name": "Justin Tremblay",
    "Age": 63,
    "Gender": "male",
    "Created": "2021-10-18T20:07:01.7496180",
    "Labels": [
        {
            "LabelId": 2
        },
        {
            "LabelId": 16
        }
    ],
    "@metadata": {
        "@collection": "Subjects",
        "Raven-Clr-Type": "ravendb_client.Model.Subjects, ravendb_client",
        "@flags": "HasTimeSeries"
    }
}
```

I created an Index "Subjects/ByLabel":
```
docs.Subjects.Select(sub => new {
    Labels = this.Recurse(sub, x => x.Labels).Select(x0 => x0.LabelId)
})
```

A query what I tested:
```
from index 'Subjects/ByLabel'
where Labels = 2 and Labels = 16
```

I wrote a simple Console Application to query against my RavenDB.

If I query against my RavenDB with that function:
```c#
public static List<Subjects> GetSubjectsByLabel_UsedIndex(int label1, int label2)
{
    using IDocumentSession session = RavenDbContext.Store.OpenSession();

    List<Subjects> subsByLabel = session
            .Query<Subjects_ByLabel.Result, Subjects_ByLabel>()
            .Where(x => x.Labels.Any(l => l == label1 && l == label2))
            .OfType<Subjects>()
            .ToList();

    return subsByLabel;
}
```

It takes about 2 seconds for a set of 10.531 documents against my 1.07m documents. But it not includes the time series, where my values are stored.

I did the approach with including the timeseries, but it takes up to 15 seconds.
```c#
IRawDocumentQuery<Subjects_ByLabel> query = session.Advanced.RawQuery<Subjects_ByLabel>
  ("from index 'Subjects/ByLabel' where Labels = $l1 and Labels = $l2 include timeseries($ts, $start, $end)")
  .AddParameter("l1", label1)
  .AddParameter("l2", label2)
  .AddParameter("start", startDate)
  .AddParameter("end", endDate)
  .AddParameter("ts", timeSeriesName);
```

And I tried to use RQL to fetch only the specific timeseries:
```c#
var baseline = new DateTime(2021, 9, 1, 00, 00, 00);
using IDocumentSession session = RavenDbContext.Store.OpenSession();

IRawDocumentQuery<TimeSeriesAggregationResult> aggregatedRawQuery =
session.Advanced.RawQuery<TimeSeriesAggregationResult>(@"
    from index 'Subjects/ByLabel'
    where Labels = 2 and Labels = 16
    select timeseries(
        from Welliba
        between $start and $end
        group by '1 days'
        select min(), max(), avg(), count()
    )
")
.AddParameter("start", baseline)
.AddParameter("end", baseline.AddMonths(1));

var aggregatedRawQueryResult = aggregatedRawQuery.ToList();
return aggregatedRawQueryResult;
```

but it takes up to 15 seconds.

My question would be, what can I improve to beat the 2 seconds for getting my data. Am I on the right track or should I reconsider things.

side note: I'm using a Docker image of RavenDB on an old MacBookPro (mid 2012, 2,6 Ghz i7, 16 GB DDR3). Will try to move that dummy database to Azure, because my machine lacks in power.

**Comment by ayende** [comment]
If I understand you properly, you have a dataset of about a million documents, and you want to query for the timeseries of specific users that match your query (based on values from the document), is that correct?

There are a few things to note here. You don't need to call `Recurse` here, that isn't what it is meant for.

The proper query can be (no index needed):

```
from Subjects
where Labels[].LableId in (1,2)
```

Now, the key issue for the duration of this query is that you are getting _all the results back_, all *1,000,000* documents.
Based on your code, you are doing that using `ToList()`, which means that both client & server will buffer the results, so the timeline is roughly:

* Prepare a buffer for 1M documents in the server side.
* Send 1M documents over the network
* Deserialize 1M documents into a `List<Subject>` on your client
* Return that list to you

Depending on the size of the documents, you are spending a _lot_ of time just passing the data around.

Your scenario actually calls for getting *all* documents? Or is it a paged scenario?
If you need all results, you should use the streaming API for this purpose.

You can also add `include timings()` to a query to see its exact performance profile on the server side.

How many values do you have per time series?  Note that for a group of `1 day`, even if you have millions of values in the time series, it should be far faster than you are seeing. I would assume that the problem is with the page size of the query.

**Reply by Welliba-Sven-Barkhahn** [reply]
My scenario would be to retrieve all documents, which match my criteria ... in my case it means I should get around 10k documents back and not the full set (1mio). Currently, It varies between 30 and 200 entries per time series per user. 

So to speak, I want to use a query that returns me the time series for the users that match my Where clause so I can then do further calculations with the data and the only way to achieve this is to use the Streaming API?

**Reply by ayende** [reply]
In that case, you should use stream, yes. See below, if you have any questions or the performance isn't what you get, let us know and we can look at this in more details.

See here for some examples:

https://ravendb.net/docs/article-page/5.2/csharp/client-api/session/querying/how-to-stream-query-results#example-ii---dynamic-document-query

https://ravendb.net/docs/article-page/5.2/csharp/client-api/session/querying/how-to-stream-query-results#example-iii---dynamic-raw-query

**Reply by Welliba-Sven-Barkhahn** [reply]
Hello @ayende,

I recognized that the query itself is fast, but the payload is more then 20 megabytes (depending on the select projection) and that's why it took so long on my client side.

Is there a way to shrink the payload, f.e. to get only the values from fields which are relevant for me, so that my documents are reduced to a minimum on information? (a lightweight version of `TimeSeriesAggregationResult` and my document to which the time series belongs)

**Reply by ayende** [reply]
How many results do you have for the query?
The output of the time series should be pretty small, just the fields that you selected.
You can do a further projection on top of that, of course, but that seems strange.

Can you show us a sample?

**Reply by Welliba-Sven-Barkhahn** [reply]
Sure I can :)

Query:
```
from Subjects 
where Labels[].LabelId all in (2, 16)
select timeseries(
    from TestSerie
    between '2021-09-01' and '2021-10-01'
    group by '1 days'
    select avg()
)
```

My query returns 5694 documents and 167457 timeseries aggregated results and the total size is 20.89 megabytes.

Single Document example:
116 Datapoints, 30 Buckets, size ~4.18 kb
[subject-example-result.json.zip](https://github.com/ravendb/ravendb/files/7424748/subject-example-result.json.zip)

**Reply by ayende** [reply]
Looking at the results, this is about 8KB formatted and about half that minified (how we send it).
Given the output that you want, what kind of details do you want _out_ of this? 
You can use something like this to strip the data further:

```
declare timeseries bpm(o)
{
    from o.HeartRates between '2020-05-01' and '2020-06-01'
    group by 5m
    select avg()
}
declare function strip(ts) {
         var n  = {};
    for(var i = 0 ; i< ts.Results.length; i++){
   
        n[ts.Results[i].From] = ts.Results[i].Average[0];
    }
    ts.Results = n;
    return ts;
}
from "Employees"  as e 
select strip(bpm(e))
```

**Reply by Welliba-Sven-Barkhahn** [reply]
Thanks for your reply.

It works when I use `session.Advanced.RawQuery<MyTestClass>(..).ToList()`.

```c#
public class MyTestClass
{
    public int Count { get; set; }
    public Dictionary<string, double> Results { get; set; }
}
```

When I tried the same Query with an Stream, I receive this exception:

Invalid JSON. Last 4KB of read stream: '{"ResultEtag":8113149660914815334,"IsStale":false,"IndexName":"Auto/Subjects/ByLabelsAndLabels[].LabelId","TotalResults":0,"IndexTimestamp":"2021-10-28T08:47:16.1560017Z","Results":[],"Error":"System.InvalidOperationException: No idea how to convert Raven.Server.Documents.Queries.Results.TimeSeries.TimeSeriesRetriever+TimeSeriesRetrieverResult to JsValue\n   at Raven.Server.Documents.Patch.JavaScriptUtils.TranslateToJs(Engine engine, JsonOperationContext context, Object o) in C:\\Builds\\RavenDB-Stable-5.2\\52012\\src\\Raven.Server\\Documents\\Patch\\JavaScriptUtils.cs:line 353\n   at Raven.Server.Documents.Patch.ScriptRunner.SingleRun.SetArgs(JsonOperationContext jsonCtx, String method, Object[] args) in C:\\Builds\\RavenDB-Stable-5.2\\52012\\src\\Raven.Server\\Documents\\Patch\\ScriptRunner.cs:line 1872\n   at Raven.Server.Documents.Patch.ScriptRunner.SingleRun.Run(JsonOperationContext jsonCtx, DocumentsOperationContext docCtx, String method, String documentId, Object[] args, QueryTimingsScope scope) in C:\\Builds\\RavenDB-Stable-5.2\\52012\\src\\Raven.Server\\Documents\\Patch\\ScriptRunner.cs:line 1865\n   at Raven.Server.Documents.Queries.Results.QueryResultRetrieverBase.InvokeFunction(String methodName, Query query, String documentId, Object[] args, QueryTimingsScope timings) in C:\\Builds\\RavenDB-Stable-5.2\\52012\\src\\Raven.Server\\Documents\\Queries\\Results\\QueryResultRetrieverBase.cs:line 862\n   at Raven.Server.Documents.Queries.Results.QueryResultRetrieverBase.GetFunctionValue(FieldToFetch fieldToFetch, String documentId, Object[] args) in C:\\Builds\\RavenDB-Stable-5.2\\52012\\src\\Raven.Server\\Documents\\Queries\\Results\\QueryResultRetrieverBase.cs:line 766\n   at Raven.Server.Documents.Queries.Results.QueryResultRetrieverBase.TryGetValue(FieldToFetch fieldToFetch, Document document, Document luceneDoc, IState state, Dictionary`2 indexFields, Nullable`1 anyDynamicIndexFields, String& key, Object& value) in C:\\Builds\\RavenDB-Stable-5.2\\52012\\src\\Raven.Server\\Documents\\Queries\\Results\\QueryResultRetrieverBase.cs:line 529\n   at Raven.Server.Documents.Queries.Results.QueryResultRetrieverBase.GetProjectionFromDocumentInternal(Document doc, Document luceneDoc, ScoreDoc scoreDoc, FieldsToFetch fieldsToFetch, JsonOperationContext context, IState state) in C:\\Builds\\RavenDB-Stable-5.2\\52012\\src\\Raven.Server\\Documents\\Queries\\Results\\QueryResultRetrieverBase.cs:line 263\n   at Raven.Server.Documents.Queries.Results.QueryResultRetrieverBase.GetProjection(Document input, ScoreDoc scoreDoc, String lowerId, IState state) in C:\\Builds\\RavenDB-Stable-5.2\\52012\\src\\Raven.Server\\Documents\\Queries\\Results\\QueryResultRetrieverBase.cs:line 138\n   at Raven.Server.Documents.Queries.Results.MapQueryResultRetriever.Get(Document input, ScoreDoc scoreDoc, IState state) in C:\\Builds\\RavenDB-Stable-5.2\\52012\\src\\Raven.Server\\Documents\\Queries\\Results\\MapQueryResultRetriever.cs:line 41\n   at Raven.Server.Documents.Indexes.Persistence.Lucene.IndexReadOperation.Query(IndexQueryServerSide query, QueryTimingsScope queryTimings, FieldsToFetch fieldsToFetch, Reference`1 totalResults, Reference`1 skippedResults, IQueryResultRetriever retriever, DocumentsOperationContext documentsContext, Func`2 getSpatialField, CancellationToken token)+MoveNext() in C:\\Builds\\RavenDB-Stable-5.2\\52012\\src\\Raven.Server\\Documents\\Indexes\\Persistence\\Lucene\\IndexReadOperation.cs:line 158\n   at Raven.Server.Utils.Enumerators.PulsedTransactionEnumerator`2.MoveNext() in C:\\Builds\\RavenDB-Stable-5.2\\52012\\src\\Raven.Server\\Utils\\Enumerators\\PulsedTransactionEnumerator.cs:line 57\n   at Raven.Server.Documents.Indexes.Index.QueryInternal[TQueryResult](TQueryResult resultToFill, IndexQueryServerSide query, QueryOperationContext queryContext, Boolean pulseDocsReadingTransaction, OperationCancelToken token) in '

**Reply by ayende** [reply]
Can you show the full code that generated this?

**Reply by ayende** [reply]
Also, what version are you running? I tested that on `live-test.ravendb.net` which is running 5.3 nightly

**Reply by Welliba-Sven-Barkhahn** [reply]
I'm using ravenDB 5.2.3 in a .Net 5.0 Console Application

```c#
namespace ravenDBClient {
  public static class RavenQueries
  {
        public class MyTimeSeriesResult
        {
            public int Count { get; set; }
            public Dictionary<string, double> Results { get; set; }
        }
        
        public static void GetSubs_WithStream(int label1, int label2, DateTime? from = null, DateTime? to = null)
        {
            using IDocumentSession session = RavenDbContext.Store.OpenSession();

            var startDate = from ?? DateTime.Today;
            var endDate = to ?? startDate.AddMonths(1);

            var query = session.Advanced.RawQuery<MyTimeSeriesResult>(@"
                declare timeseries welliba(o)
                {
                    from o.Welliba between $start and $end
                    group by '1 days'
                    select avg()
                }

                declare function strip(ts)
                {
                    var n  = {};
                    var len = ts.Results.length;
                    for(var i = 0 ; i < len; i++){
                        n[ts.Results[i].From] = ts.Results[i].Average[0];
                    }
                    ts.Results = n;
                    return ts;
                }
                from Subjects as s
                where s.Labels[].LabelId all in ($l1, $l2)
                select strip(welliba(s))
            ")
                .AddParameter("start", startDate)
                .AddParameter("end", endDate)
                .AddParameter("l1", label1)
                .AddParameter("l2", label2);
            // this works
            List<MyTimeSeriesResult> timeSeriesResults = query.ToList();
            MyTimeSeriesResult firstResult = timeSeriesResults[0];
            
            // this not
           using (var docStream = session.Advanced.Stream(query))
            {
                // throws exception here while calling MoveNext()
                while (docStream.MoveNext())
                {
                    var document = docStream.Current;
                }
            }
            
            // do stuff
       }       
   }
}
```

**Reply by ayende** [reply]
I can confirm that this is a bug, see follow up here: https://issues.hibernatingrhinos.com/issue/RavenDB-17418

The fix is here: https://github.com/ravendb/ravendb/pull/12948

**Reply by CameronProbert** [reply]
@ayende 
> Now, the key issue for the duration of this query is that you are getting _all the results back_, all _1,000,000_ documents. Based on your code, you are doing that using `ToList()`, which means that both client & server will buffer the results, so the timeline is roughly:
>     * Prepare a buffer for 1M documents in the server side.
>     * Send 1M documents over the network
>     * Deserialize 1M documents into a `List<Subject>` on your client
>     * Return that list to you

Are you saying in this answer of yours that the `Where` clause does not reduce the number of results sent from the server to the client? ToList() means they are all sent, and the `Where`  is executed on the client? In the docs it suggests that `ToList()` is how you get the query to execute. You seem to suggest there is an alternative way to get the results back though?

**Reply by ayende** [reply]
No, that isn't the case. IIRC (and that was few years ago), in that particular dataset, all documents were matched by this query

**Comment by kamranayub** [comment]
Hey @Welliba-Sven-Barkhahn 

You said:

> do some calculations with the data in the time series

Is this for calculations being displayed on the client (your app) or for administrative purposes like reporting? In general I would suggest that anything that's not user-facing be done in background jobs. Then it is less important how fast it is (though there's ways to make things fast depending on the exact use case).

You could also perform these calculations _in the index_ and then your **queries** will be even faster (at the expense of increased index CPU usage). You could do that using inline C#, Additional Sources (and even possibly custom MyGet NuGet packages).

With a better idea of what your end goal is, we can help further optimize what you need.

**Comment by CameronProbert** [comment]
@ayende 
> Now, the key issue for the duration of this query is that you are getting _all the results back_, all _1,000,000_ documents. Based on your code, you are doing that using `ToList()`, which means that both client & server will buffer the results, so the timeline is roughly:
>     * Prepare a buffer for 1M documents in the server side.
>     * Send 1M documents over the network
>     * Deserialize 1M documents into a `List<Subject>` on your client
>     * Return that list to you

Are you saying in this answer of yours that the `Where` clause does not reduce the number of results sent from the server to the client? ToList() means they are all sent, and the `Where`  is executed on the client? In the docs it suggests that `ToList()` is how you get the query to execute. You seem to suggest there is an alternative way to get the results back though?

**Reply by ayende** [reply]
No, that isn't the case. IIRC (and that was few years ago), in that particular dataset, all documents were matched by this query
